### PR TITLE
feat: Rollback + Full traceback for GraphQLErrors

### PR DESCRIPTION
**Pull request by fahimalizain**
The following is done if there is any error on graphql response:
- DB Rollback
- Make `GraphQL Error Log`
- Mask the incoming Password fields in `GraphQL Error Logs`
- Traceback in Response if `developer_mode`

<details><summary>Error Log Example</summary>

```
Query: {
    get_todo_with_user_status_sorting(first: 10) {
        totalCount
    }
    user_with_number_of_roles(first: 15, sortBy: { direction: DESC, field: NUM_ROLES }) {
        totalCount
        pageInfo {
            hasNextPage
            hasPreviousPage
            startCursor
            endCursor
        }
        edges {
            cursor
            node {
                full_name
                num_roles
                user {
                    email,
                    modified
                }
            }
        }
    }
}
Variables: None
Operation Name: None

Output:
{'data': {'get_todo_with_user_status_sorting': None, 'user_with_number_of_roles': None}, 'errors': [GraphQLError('RANDOM', locations=[SourceLocation(line=2, column=5)], path=['get_todo_with_user_status_sorting']), GraphQLError("'list' object has no attribute 'lower'", locations=[SourceLocation(line=5, column=5)], path=['user_with_number_of_roles'])]}

Tracebacks:

frappe.get_traceback():
NoneType: None

==========================================

GraphQLError tracebacks:
GQLError #0
RANDOM

GraphQL request:2:5
1 | {
2 |     get_todo_with_user_status_sorting(first: 10) {
  |     ^
3 |         totalCount

Traceback (most recent call last):
  File "/home/faztp12/dev-bench/frappe-graphql/env/lib/python3.6/site-packages/graphql/execution/execute.py", line 679, in complete_value_catching_error
    return_type, field_nodes, info, path, result
  File "/home/faztp12/dev-bench/frappe-graphql/env/lib/python3.6/site-packages/graphql/execution/execute.py", line 745, in complete_value
    raise result
  File "/home/faztp12/dev-bench/frappe-graphql/env/lib/python3.6/site-packages/graphql/execution/execute.py", line 637, in resolve_field_value_or_error
    result = resolve_fn(source, info, **args)
  File "/home/faztp12/dev-bench/frappe-graphql/apps/test_app/test_app/todo_with_user_status_sorting.py", line 35, in <lambda>
    obj, info, **kwargs),
  File "/home/faztp12/dev-bench/frappe-graphql/apps/frappe_graphql/frappe_graphql/utils/cursor_pagination.py", line 87, in resolve
    raise Exception("RANDOM")
Exception: RANDOM
==========================================



GQLError #1
'list' object has no attribute 'lower'

GraphQL request:5:5
4 |     }
5 |     user_with_number_of_roles(first: 15, sortBy: { direction: DESC, field: NUM_ROLES }) {
  |     ^
6 |         totalCount

Traceback (most recent call last):
  File "/home/faztp12/dev-bench/frappe-graphql/env/lib/python3.6/site-packages/graphql/execution/execute.py", line 679, in complete_value_catching_error
    return_type, field_nodes, info, path, result
  File "/home/faztp12/dev-bench/frappe-graphql/env/lib/python3.6/site-packages/graphql/execution/execute.py", line 745, in complete_value
    raise result
  File "/home/faztp12/dev-bench/frappe-graphql/env/lib/python3.6/site-packages/graphql/execution/execute.py", line 637, in resolve_field_value_or_error
    result = resolve_fn(source, info, **args)
  File "/home/faztp12/dev-bench/frappe-graphql/apps/test_app/test_app/user_with_number_of_roles.py", line 101, in <lambda>
    ).resolve(obj, info, **kwargs)
  File "/home/faztp12/dev-bench/frappe-graphql/apps/frappe_graphql/frappe_graphql/utils/cursor_pagination.py", line 43, in resolve
    sort_key, sort_dir = self.get_sort_args(kwargs.get("sortBy"))
  File "/home/faztp12/dev-bench/frappe-graphql/apps/frappe_graphql/frappe_graphql/utils/cursor_pagination.py", line 151, in get_sort_args
    sort_key = sorting_input.get("field").lower()
AttributeError: 'list' object has no attribute 'lower'
==========================================
```
</details>

<details><summary>Traceback in Response Example</summary>

```json
{
    "data": {
        "get_todo_with_user_status_sorting": null,
        "user_with_number_of_roles": null
    },
    "errors": [
        "GraphQLError('RANDOM', locations=[SourceLocation(line=2, column=5)], path=['get_todo_with_user_status_sorting'])",
        "GraphQLError(\"'list' object has no attribute 'lower'\", locations=[SourceLocation(line=5, column=5)], path=['user_with_number_of_roles'])"
    ],
    "exc": "[\"frappe.get_traceback: NoneType: None\\n\", \"['GQLError #0\\\\nRANDOM\\\\n\\\\nGraphQL request:2:5\\\\n1 | {\\\\n2 |     get_todo_with_user_status_sorting(first: 10) {\\\\n  |     ^\\\\n3 |         totalCount\\\\n\\\\nTraceback (most recent call last):\\\\n  File \\\"/home/faztp12/dev-bench/frappe-graphql/env/lib/python3.6/site-packages/graphql/execution/execute.py\\\", line 679, in complete_value_catching_error\\\\n    return_type, field_nodes, info, path, result\\\\n  File \\\"/home/faztp12/dev-bench/frappe-graphql/env/lib/python3.6/site-packages/graphql/execution/execute.py\\\", line 745, in complete_value\\\\n    raise result\\\\n  File \\\"/home/faztp12/dev-bench/frappe-graphql/env/lib/python3.6/site-packages/graphql/execution/execute.py\\\", line 637, in resolve_field_value_or_error\\\\n    result = resolve_fn(source, info, **args)\\\\n  File \\\"/home/faztp12/dev-bench/frappe-graphql/apps/test_app/test_app/todo_with_user_status_sorting.py\\\", line 35, in <lambda>\\\\n    obj, info, **kwargs),\\\\n  File \\\"/home/faztp12/dev-bench/frappe-graphql/apps/frappe_graphql/frappe_graphql/utils/cursor_pagination.py\\\", line 87, in resolve\\\\n    raise Exception(\\\"RANDOM\\\")\\\\nException: RANDOM\\\\n==========================================\\\\n\\\\n', 'GQLError #1\\\\n\\\\'list\\\\' object has no attribute \\\\'lower\\\\'\\\\n\\\\nGraphQL request:5:5\\\\n4 |     }\\\\n5 |     user_with_number_of_roles(first: 15, sortBy: { direction: DESC, field: NUM_ROLES }) {\\\\n  |     ^\\\\n6 |         totalCount\\\\n\\\\nTraceback (most recent call last):\\\\n  File \\\"/home/faztp12/dev-bench/frappe-graphql/env/lib/python3.6/site-packages/graphql/execution/execute.py\\\", line 679, in complete_value_catching_error\\\\n    return_type, field_nodes, info, path, result\\\\n  File \\\"/home/faztp12/dev-bench/frappe-graphql/env/lib/python3.6/site-packages/graphql/execution/execute.py\\\", line 745, in complete_value\\\\n    raise result\\\\n  File \\\"/home/faztp12/dev-bench/frappe-graphql/env/lib/python3.6/site-packages/graphql/execution/execute.py\\\", line 637, in resolve_field_value_or_error\\\\n    result = resolve_fn(source, info, **args)\\\\n  File \\\"/home/faztp12/dev-bench/frappe-graphql/apps/test_app/test_app/user_with_number_of_roles.py\\\", line 101, in <lambda>\\\\n    ).resolve(obj, info, **kwargs)\\\\n  File \\\"/home/faztp12/dev-bench/frappe-graphql/apps/frappe_graphql/frappe_graphql/utils/cursor_pagination.py\\\", line 43, in resolve\\\\n    sort_key, sort_dir = self.get_sort_args(kwargs.get(\\\"sortBy\\\"))\\\\n  File \\\"/home/faztp12/dev-bench/frappe-graphql/apps/frappe_graphql/frappe_graphql/utils/cursor_pagination.py\\\", line 151, in get_sort_args\\\\n    sort_key = sorting_input.get(\\\"field\\\").lower()\\\\nAttributeError: \\\\'list\\\\' object has no attribute \\\\'lower\\\\'\\\\n==========================================\\\\n\\\\n']\"]"
}
```
</details>

### CODE REVIEW

File: frappe_graphql/api.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        query=query,
+        variables=variables,
+        operation_name=operation_name
+    )
+
+    frappe.local.response = output
+    if len(output.get("errors", [])):
+        frappe.db.rollback()
+        log_error(query, variables, operation_name, output)
+        frappe.local.response["http_status_code"] = 400
```
Comment (e-lobo) [inline frappe_graphql/api.py]: extract from exception thrown is better right ? @assemmarwan ? frappe.throw default is 
image

Comment (fahimalizain) [reply frappe_graphql/api.py]: ok. i will fetch the max that was thrown

**Comment by assemmarwan** [comment]
@faztp12 This PR changes the structure of the response which breaks the GraphQL client in the frontend 😓 

Then:

```json
{
  "data": {
    "me": {
      "user": null
    }
  },
  "_debug_messages": "[\"method:frappe_graphql.graphql.execute\"]"
}
```

Now: 

```json
{
  "message": {  // <======================
    "data": {
      "me": {
        "user": null
      }
    }
  },
  "_debug_messages": "[\"method:frappe_graphql.graphql.execute\"]"
}
```

**Comment by fahimalizain** [comment]
@assemmarwan I think it is the cache. Please clear your cache to get hooks updated ?
The method listed in `_debug_messages` is the old one.

image

File: frappe_graphql/api.py (hunk 1 above)
Comment (fahimalizain) [reply frappe_graphql/api.py]: @assemmarwan @e-lobo updated

**Comment by assemmarwan** [comment]
What do you think of having a setting (maybe under Systems Settings) to toggle logging?

File: frappe_graphql/api.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+def execute_gql_query():
+    query, variables, operation_name = get_query()
+    output = execute(
+        query=query,
+        variables=variables,
+        operation_name=operation_name
+    )
+
+    frappe.clear_messages()
+    frappe.local.response = output
```
Comment (assemmarwan) [inline frappe_graphql/api.py:20]: @faztp12 What's the difference between returning output vs setting it to local response like here?

File: frappe_graphql/utils/http.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    """
+    Gets the active operation name
+    if operation_name is not specified in the request,
+    it will take on the first operation definition if available.
+    Otherwise returns 'unnamed-query'
+    """
+    defined_operations = []
+    try:
+        document = parse(query)
+        print(document)
```
Comment (assemmarwan) [inline frappe_graphql/utils/http.py]: @faztp12 a wild print statement 😛

File: frappe_graphql/api.py (hunk 2 above)
Comment (fahimalizain) [reply frappe_graphql/api.py:20]: @assemmarwan when output is returned, it will be wrapped in `message`

Comment (assemmarwan) [reply frappe_graphql/api.py:20]: Thanks!

**Review by assemmarwan — APPROVED** [review]

File: frappe_graphql/utils/http.py (hunk 3 above)
Comment (fahimalizain) [reply frappe_graphql/utils/http.py]: @assemmarwan fixed :D

**Review by e-lobo — APPROVED** [review]

**Pull request merged**